core/src/main/java/org/apache/iceberg/hadoop/HadoopInputFile.java

```diff
@@ -54,51 +55,60 @@ public static HadoopInputFile fromLocation(CharSequence location, long length,
     return fromPath(path, length, conf);
   }
 
+  public static HadoopInputFile fromLocation(CharSequence location, FileSystem fs) {
+    Path path = new Path(location.toString());
+    return fromPath(path, fs);
+  }
+
+  public static HadoopInputFile fromLocation(CharSequence location, long length,
+                                             FileSystem fs) {
+    Path path = new Path(location.toString());
+    return fromPath(path, length, fs);
+  }
+
   public static HadoopInputFile fromPath(Path path, Configuration conf) {
-    try {
-      FileSystem fs = path.getFileSystem(conf);
-      return new HadoopInputFile(fs, path, conf);
-    } catch (IOException e) {
-      throw new RuntimeIOException(e, "Failed to get file system for path: %s", path);
-    }
+    FileSystem fs = Util.getFs(path, conf);
+    return fromPath(path, fs);
   }
 
   public static HadoopInputFile fromPath(Path path, long length, Configuration conf) {
-    try {
-      FileSystem fs = path.getFileSystem(conf);
-      return new HadoopInputFile(fs, path, length, conf);
-    } catch (IOException e) {
-      throw new RuntimeIOException(e, "Failed to get file system for path: %s", path);
-    }
+    FileSystem fs = Util.getFs(path, conf);
+    return fromPath(path, length, fs);
+  }
+
+  public static HadoopInputFile fromPath(Path path, FileSystem fs) {
+    return new HadoopInputFile(fs, path);
+  }
+
+  public static HadoopInputFile fromPath(Path path, long length, FileSystem fs) {
+    return new HadoopInputFile(fs, path, length);
   }
 
   public static HadoopInputFile fromStatus(FileStatus stat, Configuration conf) {
-    try {
-      FileSystem fs = stat.getPath().getFileSystem(conf);
-      return new HadoopInputFile(fs, stat, conf);
-    } catch (IOException e) {
-      throw new RuntimeIOException(e, "Failed to get file system for path: %s", stat.getPath());
-    }
+    FileSystem fs = Util.getFs(stat.getPath(), conf);
+    return fromStatus(stat, fs);
+  }
+
+  public static HadoopInputFile fromStatus(FileStatus stat, FileSystem fs) {
+    return new HadoopInputFile(fs, stat);
   }
 
-  private HadoopInputFile(FileSystem fs, Path path, Configuration conf) {
+  private HadoopInputFile(FileSystem fs, Path path) {
     this.fs = fs;
     this.path = path;
-    this.conf = conf;
   }
 
-  private HadoopInputFile(FileSystem fs, Path path, long length, Configuration conf) {
+  private HadoopInputFile(FileSystem fs, Path path, long length) {
+    checkArgument(length >= 0);
```

Comment:
This should include an error message that explains what went wrong, like `"Invalid input file length: %s"`